types/validator_set.go

```diff
@@ -48,12 +48,22 @@ func NewValidatorSet(vals []*Validator) *ValidatorSet {
 }
 
 // incrementAccum and update the proposer
-// TODO: mind the overflow when times and votingPower shares too large.
 func (valSet *ValidatorSet) IncrementAccum(times int) {
 	// Add VotingPower * times to each validator and order into heap.
 	validatorsHeap := cmn.NewHeap()
 	for _, val := range valSet.Validators {
-		val.Accum += val.VotingPower * int64(times) // TODO: mind overflow
+		// check for overflow both multiplication and sum
+		res, overflow := safeMul(val.VotingPower, int64(times))
```

Comment:
maybe we can simplify this somehow?

Reply:
safeMulClip could automatically set to math.MaxInt64 or math.MinInt64 appropriately, and return a single value.  safeAddClip likewise.